lib/utils.js

```diff
@@ -966,3 +966,32 @@ exports.errorToPOJO = function errorToPOJO(error) {
 exports.warn = function warn(message) {
   return process.emitWarning(message, { code: 'MONGOOSE' });
 };
+
+
+exports.isCircular = function isCircular(obj) {
+  const seenObjectsMap = new Map();
+
+  return detect(obj, '');
+
+  function detect(obj, basePath) {
+    if (obj && typeof obj === 'object') {
+      if (seenObjectsMap.get(obj)) {
+        return { isCircular: true, path: basePath };
+      }
+      seenObjectsMap.set(obj, true);
```

Comment:
if you set everything to true, then you could actually just use  a Set and not a Map. Probably a WeakSet.